libs/src/service-selection/ServiceSelection.js

```diff
@@ -0,0 +1,223 @@
+const _ = require('lodash')
+const { raceRequests } = require('../utils/network')
+
+/**
+ * A class that assists with autoselecting services.
+ * `ServiceSelection` is intended to be overridden with further
+ * business logic that a particular sevice might preference.
+ *
+ * The general use case is as follows:
+ *
+ * ```
+ *
+ * const selector = new ServiceSelection({
+ *   getServices: ethContracts.getDiscoveryProviders()
+ * })
+ *
+ * const service = await selector.select()
+ *
+ * ```
+ *
+ * This class operates by taking a list of services and
+ * round-robin makes requests at them until a suitable one is found.
+ *
+ * Two types of "bad" services are defined below:
+ *  - Unhealthy: this service is bad and should not be used
+ *  - Backup: this service is bad, but if we can't find anything better, maybe use it
+ *
+ * Classes that extend `ServiceSelection` can choose to implement custom logic on top
+ * of them and is generally how this class is intended to be used.
+ */
+class ServiceSelection {
+  /**
+   * @param config
+   * @param {Set<string>} config.whitelist only services from this list are allowed to be picked
+   * @param {() => Promise<String[]>} conffig.getServices an (async) method to get a list of services
```

Comment:
small typo at `conffig` :)

Reply:
fixed in the second PR which i'll merge into this